### PR TITLE
📝 Add documentation about compering old and new version of snapshot

### PR DESCRIPTION
**Pull request by dawidreedsy**
As per this discussion:
- https://github.com/share/sharedb/pull/667#discussion_r1654876315

### CODE REVIEW

File: docs/middleware/op-submission.md [hunk 1]
```diff
@@ -157,3 +157,21 @@ backend.use('apply', (request, next) => {
...
 {: .warn :}
 The `request.$fixup()` method may throw an error, which should be handled appropriately, usually by passing directly to the `next()` callback.
+
+## Comparing old snapshot version with new version
+
+Frequently, it becomes necessary to verify the changes made. This can be accomplished by leveraging two hooks, `apply` and `commit`, and creating a snapshot clone within the `apply` hook.
+
+```js
+backend.use('apply', (request, next) => {
+  request.snapshotBeforeApply = utils.clone(request.snapshot);
```
Comment (alecgibson) [inline docs/middleware/op-submission.md:167]: I don't think we export `utils`, so should probably either:

 - change this to a generic `clone()` with a note about picking your favourite deep clone; or
 - just replace with `JSON.parse(JSON.stringify())`, which is [what this does anyway](https://github.com/share/sharedb/blob/dcef5491ca400826389c823a9ae2c53e2b02beac/lib/util.js#L98)

File: docs/middleware/op-submission.md [hunk 2]
```diff
@@ -157,3 +157,21 @@ backend.use('apply', (request, next) => {
...
+  next(error);
+});
+
+backend.use('commit', (request, next) => {
+  // Snapshot without ops and $fixupOps applied is now available as snapshotBeforeApply
+  // Snapshot with ops and $fixupOps applied is still available as snapshot
```
Comment (alecgibson) [inline docs/middleware/op-submission.md:173]:
```suggestion
  // Snapshot with ops and $fixupOps applied is still available as request.snapshot
```

File: docs/middleware/op-submission.md [hunk 3]
```diff
@@ -157,3 +157,24 @@ backend.use('apply', (request, next) => {
...
+
+## Comparing old snapshot version with new version
+
+Frequently, it becomes necessary to verify the changes made. This can be accomplished by leveraging two hooks, `apply` and `commit`, and creating a snapshot clone within the `apply` hook.
+
+```js
+backend.use('apply', (request, next) => {
+  // Here use your favorite deep clone implementation
+  // sharedb team recommends rfdc
+  // https://github.com/davidmarkclements/rfdc#readme
```
Comment (alecgibson) [inline docs/middleware/op-submission.md:169]: I think let's skip the recommendation here; `sharedb` doesn't actually use this internally anywhere.

File: docs/middleware/op-submission.md [hunk 4]
```diff
@@ -157,3 +157,27 @@ backend.use('apply', (request, next) => {
...
 {: .warn :}
 The `request.$fixup()` method may throw an error, which should be handled appropriately, usually by passing directly to the `next()` callback.
+
+## Comparing old snapshot version with new version
+
+Frequently, it becomes necessary to verify the changes made. This can be accomplished by leveraging two hooks, `apply` and `commit`, and creating a snapshot clone within the `apply` hook.
+
+```js
+// This is simple function that can deeply clone the object
+// Note that this function is not most efficient one
```
Comment (alecgibson) [inline docs/middleware/op-submission.md:167]: I think no need to add commentary on efficiency; as @ericyhwang points out, it often depends on the exact shape of the thing you're cloning, and I don't think it's our place to determine this for people.

```suggestion
// Alternatively use your favourite deep-clone library
```

**Review by alecgibson — APPROVED** [review]

**Review by ericyhwang — APPROVED** [review]

**Pull request merged**